### ISSUE TITLE
Add CO2 as a requested variable for TM5

### ISSUE THREAD

**Issue by nierad** [EC-Earth/ece2cmor3#589]
Hi all, 

I was about to test the TM5 CO2-only setup with Tommi's extra CO2 output. I then quickly added CO2 to the Amon table, which made TM5 produce the output but ece2cmor3 didn't want to produce a cmorized file from it. 

The error message is:
`ERROR:ece2cmor3.taskloader: Variable co2 in table Amon is not supported by tm5 in ece2cmor3; if you do expect an ec-earth output variable here, please create an issue or pull request on our github page. 
`

This is hereby done:)

I am not sure though, whether it should be an Amon or another table's entry.

**Comment by etiennesky** [comment]
co2 (mol mol-1) and co2mass (kg) are requested in the Amon tables.  Would it be possible to have not only co2 but co2mass cmorized as well?

**Comment by etiennesky** [comment]
See issue on the ec-earth portal : https://dev.ec-earth.org/issues/706

**Comment by treerink** [comment]
Hi I need to know what we are talking about: Which MIP(s), which experiment(s), which table-variable combination(s). 

From some reading I guess you are asking for C4MIP? Other MIPs as well?

I had a look in the ignored list of ece2cmor, and find the following `co2` & `co2mass` cases (see far right for MIP info and previous identification log):

```
table   variable  prio dimensions                       long_name                      unit       list of MIPs which request this variable                                                                comment_author        comment

Amon    co2       1    longitude latitude plev19 time   Mole Fraction of CO2           mol mol-1  AerChemMIP,C4MIP,CFMIP,CMIP,DAMIP,FAFMIP,GMMIP,GeoMIP,HighResMIP,LS3MIP,LUMIP,PMIP,RFMIP,VolMIP         Tommi Bergman, Thomas  Not available in TM5. CO2 only transported in C-Cycle version, unclear if it is needed as output from TM5? 
Amon    co2mass   1    time                             Total Atmospheric Mass of CO2  kg         AerChemMIP,C4MIP,CDRMIP,CFMIP,CMIP,DAMIP,FAFMIP,GMMIP,GeoMIP,HighResMIP,LS3MIP,LUMIP,PMIP,RFMIP,VolMIP  Tommi Bergman, Thomas  Not available in TM5. 

AERmon  co2       1    longitude latitude alevel time   Mole Fraction of CO2           mol mol-1  AerChemMIP,DAMIP                                                                                        Tommi Bergman, Thomas  Not available in TM5. CO2 only transported in C-Cycle version, unclear if it is needed as output from TM5? 

E3hrPt  co2       1    longitude latitude alevel time1  Mole Fraction of CO2           mol mol-1  RFMIP                                                                                                   Tommi Bergman, Thomas  Not available in TM5. CO2 only transported in C-Cycle version, unclear if it is needed as output from TM5? 
```
I guess only the first two are relevant?

**Comment by nierad** [comment]
Hi @treerink,

It is definitely C4MIP and I think it is the only one. Isn't the list of experiments given by the original CMIP6 data-request? 
My guess would be to add it to all CMIP/EC-Earth-CC/... experiments. @etiennesky what do you think? 

Re Variables: Definitely the first two. I don't know if anyone wants to use the other ones.

**Comment by treerink** [comment]
Do I conclude correctly that both `co2` & `co2mass` are now outputted by TM5 only in the `r7253_co2_output_tm5` branch? Is this coming to the trunk and to the TM5 version in EC-Earth3.3.3? As long it is in a branch we have to agree how and where to provide the updated control-output files.

**Comment by etiennesky** [comment]
Hi so we need to output co2 and co2mass from Amon for the c4mip runs, but only the esm- variants which are esm-picontrol, esm-hist and esm-ssp585 .

In practical terms, if it could be added to the following defs it would be fine:

CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-esm-piControl
CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-esm-hist
C4MIP/cmip6-experiment-C4MIP-esm-ssp585

I am unsure why co2 is not output by the aerchemip runs, since it seems to be requested. But maybe @tommibergman or @plesager can respond.

@treerink  co2 and co2mass are currently only in the dev branch  r7253_co2_output_tm5 but we need it into trunk for the next release.

**Comment by tommibergman** [comment]
CO2 is not outputted for AerChemMIP, since  TM5 is not used to transport CO2 in that version.

**Comment by tommibergman** [comment]
@etiennesky @treerink If I recall correctly co2mass is not implemented, but only co2 mixing ratio. Easiest would be to add the mass/m2 and postprocess it to global sum.

**Comment by etiennesky** [comment]
> @etiennesky @treerink If I recall correctly co2mass is not implemented, but only co2 mixing ratio. Easiest would be to add the mass/m2 and postprocess it to global sum.

My understanding is that we can compute it using ece2cmor. is there another tm5 variable for which this is done?

**Comment by treerink** [comment]
@tommibergman I try to understand your `co2mass` suggestion. Could you in terms of TM5 cmor variables give the  expression for `co2mass`then?

**Comment by tommibergman** [comment]
Well it is mol(co2)/mol(air), which would require to have the airmass also from TM5. It would be easier to just output co2mass separately from TM5 after that it is just sum(areacella(TM5)*co2mass) in ece2cmor3. co2mass(in kg/m2) can be added easily to current output.

**Comment by etiennesky** [comment]
@tommibergman that would be great and an easy way to diagnose things. Note that the data request is for monthly mean. Could you also add it as daily so we can easily have the value on a given date(e.g. first/last day of year)?

**Comment by treerink** [comment]
@tommibergman  Do I understand correctly that you will add `co2mass` and `areacella` in the TM5 output in the r7253_co2_output_tm5 branch before it will be merged into the trunk? And that I include `co2` & `co2mass` into `ece2cmor3/genecec`. Where  `co2mass` =  `sum(areacella_TM5` * `co2mass)` which thereafter is converted to kg/m2 by `ece2cmor3`?

At which frequency is the TM5 output for these two variables afterall?

I guess the `areacella` can be computed as well in `ece2cmor3` by using `cdo gridarea any-tm5-file.nc` like [here](https://dev.ec-earth.org/issues/748#note-3).

**Comment by tommibergman** [comment]
areacella is already in the TM5 output. I already started on co2mass work, and I plan to get it working today.

**Comment by tommibergman** [comment]
But I will have an operation on my back on Friday, therefore work on the ece2cmor I cannot do in any short time scales. Someone else has to step in or the cmorisation has to wait.

**Comment by etiennesky** [comment]
Hi @tommibergman hope everything goes well.   You will output co2mass for every gridcell, then ece2cmor3 has to compute the global sum, but without the vertical interpolation to plev19 required for co2? 

Would it be easy to output  the global sum of co2mass from tm5 instead? This would make the output files much smaller and avoid more post-processing.

We can take care of things on the cmor side once the proper code is in place in tm5. Thanks for the help @tommibergman .

**Comment by etiennesky** [comment]
co2 and co2mass are correctly output on hr, daily and monthly scales. I haven't checked if co2mass is ok, but co2 is fine and just like ch4 (same units, etc), but 10 levels instead of 34. @aearamos will test cmorization - @treerink can you help out the implementation?

**Comment by etiennesky** [comment]
I was able to compute co2mass from files generated by tm5 (co2mass, areacella) by using the following commands 

```
cdo mul co2mass_AERmon_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001-185012.nc areacella_AERfx_EC-Earth3-CC_a2jk_r1i1p1f1_gn.nc tmp1.nc
cdo fldsum tmp1.nc tmp2.nc
```
The resulting file has the proper dimensions & values.

@treerink How do we do this in ece2cmor3? Thanks

**Comment by etiennesky** [comment]
Would we need to add code similar to this in tm52cmor.py but detecting that the variable is co2mass and do an area-weighted sum instead of mean?

```
    elif task.target.dims==0:
        # global means                                                                                                                                                                                                                          
        missval = getattr(ncvar,"missing_value",getattr(ncvar,"_FillValue",numpy.nan))
        vals=numpy.copy(ncvar[:])
        vals=numpy.mean(vals,axis=(1))

        # calculate area-weighted mean                                                                                                                                                                                                          
        vals=numpy.sum((vals*areacella_[numpy.newaxis,:,:]),axis=(1,2))/numpy.sum(areacella_)
```

**Comment by treerink** [comment]
The `add-co2+co2mass` branch is added. 

@etiennesky could you add your [suggestion above ](https://github.com/EC-Earth/ece2cmor3/issues/589#issuecomment-599702357) in the  `add-co2+co2mass` branch?

**Comment by etiennesky** [comment]
Hi @treerink thanks for the help - I am currently testing this

**Comment by etiennesky** [comment]
I pushed a fix, to the  add-co2+co2mass and found a bug with the area-weighted mean (used for ch4global), it was missing this statement: `ncvar=vals.copy()`

I did not test with ch4global, @plesager can you test that with this fix?

**Comment by treerink** [comment]
@etiennesky, so if I understand correctly you don't need an expression for `co2mass` with your current implementation. I wonder whether whether for `co2mass` the `source` in the `tm5par.json` is used, currently it is set to `co2mass,` I thought it might need to be set to `co2`, but I have the impression now you are not using the `source` info of `co2mass`?

**Comment by etiennesky** [comment]
Hi @treerink I am using the co2mass files directly generated by tm5, it has using kg/m2 and a 2d grid. I don't compute from the co2 output by tm5.

Here are the files currently generated by tm5 with the co2-tm5 EC-Earth branch:

```
areacella_AERfx_EC-Earth3-CC_a2jk_r1i1p1f1_gn.nc                      co2mass_AERday_EC-Earth3-CC_a2jk_r1i1p1f1_gn_18500101-18501231.nc         ps_AERday_EC-Earth3-CC_a2jk_r1i1p1f1_gn_18500101-18501231.nc
co2_AERday_EC-Earth3-CC_a2jk_r1i1p1f1_gn_18500101-18501231.nc         co2mass_AERhr_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001010000-185012312359.nc  ps_AERhr_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001010000-185012312359.nc
co2_AERhr_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001010000-185012312359.nc  co2mass_AERmon_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001-185012.nc             ps_AERmon_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001-185012.nc
co2_AERmon_EC-Earth3-CC_a2jk_r1i1p1f1_gn_185001-185012.nc             orog_AERfx_EC-Earth3-CC_a2jk_r1i1p1f1_gn.nc                               sftlf_AERfx_EC-Earth3-CC_a2jk_r1i1p1f1_gn.nc
```

**Comment by treerink** [comment]
> Hi @treerink I am using the co2mass files directly generated by tm5

Ok perfect, than the tm5par.json is fine as it is in this branch now.

**Comment by treerink** [comment]
Testing this branch and running `genecec` with it.

**Comment by treerink** [comment]
After running `genecec` I get the addition of  `co2` & `co2mass` in the following `json data request` files (the file names show the experiments):
```
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-hist-1950HC/cmip6-data-request-varlist-AerChemMIP-hist-1950HC-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-hist-piNTCF/cmip6-data-request-varlist-AerChemMIP-hist-piNTCF-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-histSST/cmip6-data-request-varlist-AerChemMIP-histSST-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-histSST-1950HC/cmip6-data-request-varlist-AerChemMIP-histSST-1950HC-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-histSST-piCH4/cmip6-data-request-varlist-AerChemMIP-histSST-piCH4-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-histSST-piNTCF/cmip6-data-request-varlist-AerChemMIP-histSST-piNTCF-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-piClim-2xdust/cmip6-data-request-varlist-AerChemMIP-piClim-2xdust-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-piClim-CH4/cmip6-data-request-varlist-AerChemMIP-piClim-CH4-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-piClim-HC/cmip6-data-request-varlist-AerChemMIP-piClim-HC-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-piClim-NTCF/cmip6-data-request-varlist-AerChemMIP-piClim-NTCF-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-ssp370-lowNTCF/cmip6-data-request-varlist-AerChemMIP-ssp370-lowNTCF-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-ssp370SST/cmip6-data-request-varlist-AerChemMIP-ssp370SST-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-ssp370SST-lowCH4/cmip6-data-request-varlist-AerChemMIP-ssp370SST-lowCH4-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-ssp370SST-lowNTCF/cmip6-data-request-varlist-AerChemMIP-ssp370SST-lowNTCF-EC-EARTH-AerChem.json
cmip6-output-control-files/AerChemMIP/cmip6-experiment-AerChemMIP-ssp370SST-ssp126Lu/cmip6-data-request-varlist-AerChemMIP-ssp370SST-ssp126Lu-EC-EARTH-AerChem.json
cmip6-output-control-files/C4MIP/cmip6-experiment-C4MIP-1pctCO2-bgc/cmip6-data-request-varlist-C4MIP-1pctCO2-bgc-EC-EARTH-CC.json
cmip6-output-control-files/C4MIP/cmip6-experiment-C4MIP-esm-ssp585/cmip6-data-request-varlist-C4MIP-esm-ssp585-EC-EARTH-CC.json
cmip6-output-control-files/CDRMIP/cmip6-experiment-CDRMIP-1pctCO2-cdr/cmip6-data-request-varlist-CDRMIP-1pctCO2-cdr-EC-EARTH-CC.json
cmip6-output-control-files/CDRMIP/cmip6-experiment-CDRMIP-esm-pi-cdr-pulse/cmip6-data-request-varlist-CDRMIP-esm-pi-cdr-pulse-EC-EARTH-CC.json
cmip6-output-control-files/CDRMIP/cmip6-experiment-CDRMIP-esm-pi-CO2pulse/cmip6-data-request-varlist-CDRMIP-esm-pi-CO2pulse-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-1pctCO2/cmip6-data-request-varlist-CMIP-1pctCO2-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-abrupt-4xCO2/cmip6-data-request-varlist-CMIP-abrupt-4xCO2-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-amip/cmip6-data-request-varlist-CMIP-amip-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-esm-hist/cmip6-data-request-varlist-CMIP-esm-hist-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-esm-piControl/cmip6-data-request-varlist-CMIP-esm-piControl-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-historical/cmip6-data-request-varlist-CMIP-historical-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-AerChem/cmip6-experiment-CMIP-piControl/cmip6-data-request-varlist-CMIP-piControl-EC-EARTH-AerChem.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-1pctCO2/cmip6-data-request-varlist-CMIP-1pctCO2-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-abrupt-4xCO2/cmip6-data-request-varlist-CMIP-abrupt-4xCO2-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-amip/cmip6-data-request-varlist-CMIP-amip-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-esm-hist/cmip6-data-request-varlist-CMIP-esm-hist-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-esm-piControl/cmip6-data-request-varlist-CMIP-esm-piControl-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-historical/cmip6-data-request-varlist-CMIP-historical-EC-EARTH-CC.json
cmip6-output-control-files/CMIP/EC-EARTH-CC/cmip6-experiment-CMIP-piControl/cmip6-data-request-varlist-CMIP-piControl-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-deforest-globe/cmip6-data-request-varlist-LUMIP-deforest-globe-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-esm-ssp585-ssp126Lu/cmip6-data-request-varlist-LUMIP-esm-ssp585-ssp126Lu-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-hist-noLu/cmip6-data-request-varlist-LUMIP-hist-noLu-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-land-hist-altStartYear/cmip6-data-request-varlist-LUMIP-land-hist-altStartYear-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-land-noLu/cmip6-data-request-varlist-LUMIP-land-noLu-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-ssp126-ssp370Lu/cmip6-data-request-varlist-LUMIP-ssp126-ssp370Lu-EC-EARTH-CC.json
cmip6-output-control-files/LUMIP/cmip6-experiment-LUMIP-ssp370-ssp126Lu/cmip6-data-request-varlist-LUMIP-ssp370-ssp126Lu-EC-EARTH-CC.json
cmip6-output-control-files/ScenarioMIP/EC-EARTH-AerChem/cmip6-experiment-ScenarioMIP-ssp126/cmip6-data-request-varlist-ScenarioMIP-ssp126-EC-EARTH-AerChem.json
cmip6-output-control-files/ScenarioMIP/EC-EARTH-AerChem/cmip6-experiment-ScenarioMIP-ssp245/cmip6-data-request-varlist-ScenarioMIP-ssp245-EC-EARTH-AerChem.json
cmip6-output-control-files/ScenarioMIP/EC-EARTH-AerChem/cmip6-experiment-ScenarioMIP-ssp370/cmip6-data-request-varlist-ScenarioMIP-ssp370-EC-EARTH-AerChem.json
cmip6-output-control-files/ScenarioMIP/EC-EARTH-AerChem/cmip6-experiment-ScenarioMIP-ssp585/cmip6-data-request-varlist-ScenarioMIP-ssp585-EC-EARTH-AerChem.json
```

**Comment by plesager** [comment]
Should not appear in EC-EARTH-AerChem or AerChemMIP,  because co2 is not available in EC-Earth3-AerChem. Maybe in the future when we extend the coupling between LPJG-TM5 to EC-Earth3-AerChem.
Or is the purpose to have it, but to mute it? @noije?

**Comment by etiennesky** [comment]
maybe it's not worth the effort to exclude it? In any case, ece2cmor3 will just warn that the co2 and co2mass variables are not available. Get that with ch4 when cmorizing EC-Earth-CC output.

**Comment by noije** [comment]
I agree with you, Philippe. We don't need this CO2 output in EC-Earth3-AerChem.

**Comment by noije** [comment]
Didn't see your last comment, Etienne. If it doesn't cause any problems, it's fine.

**Comment by etiennesky** [comment]
If we want to be precise, then we cannot output ch4 and ch4global in the EC-Earth-CC experiments.

And co2, co2mass are only available in the EC-Earth-CC -esm ones i.e. CMIP-esm-hist / CMIP-esm-piControl / C4MIP-esm-ssp585 .

But I am not picky on this, I don't mind having the warnings.

Up to you guys.

**Comment by plesager** [comment]
I do not mind the warnings either. Let's keep it open for a possible future extension then.

**Comment by treerink** [comment]
Okay all added in the `r7597-control-output-files` branch in the ec-earth repository.

**Comment by treerink** [comment]
Done, closing.
